cabal-testsuite/Test/Cabal/Prelude.hs

```diff
@@ -573,6 +578,31 @@ hasSharedLibraries = do
     shared_libs_were_removed <- ghcVersionIs (>= mkVersion [7,8])
     return (not (buildOS == Windows && shared_libs_were_removed))
 
+-- | Check if the GHC that is used for compiling package tests has
+-- a shared library of the cabal library under test in its database.
+--
+-- An example where this is needed is if you want to dynamically link
+-- detailed-0.9 test suites, since those depend on the Cabal library unde rtest.
+hasCabalShared :: TestM Bool
+hasCabalShared = do
```

Comment:
This is an interesting implementation of `hasCabalShared`. At the very least, I would add a note that it is a relatively expensive check, because it involves banging on GHC, so you really only want to call this once, if at all possible.

Assuming that this is the best way to go about doing this, I have some comments:

* First, are you sure that `runProgramM ghcProgram` isn't enough to invoke GHC?  I believe that GHC should already be configured in `testProgramDb`; the configuration was done in `Test/Cabal/Monad.hs` so I don't think we should have to do it again. If this doesn't work, there is probably a different bug somewhere. Is the problem that `runM` calls `requireSuccess` whereas you need to get a boolean out instead? Maybe we can refactor those functions so you get a version that raises exceptions, and one that doesn't. This is related to #4292 where we wanted to handle `fails` more systematically.

* Second, the invocation of GHC doesn't attempt to control what packages are in scope, which means that you could end up in a situation where there are two `Cabal` packages in scope. I guess GHC is preferring the latest version so it's working correctly in this case, but it would be more robust to explicitly specify which Cabal library we want to use, and the test would stop working the moment someone uses a GHC that comes with a newer version of Cabal.

Is this check the way we want to go about it? I'll admit, I can't think of very many other good choices. One possibility is to look at the `LocalBuildInfo` when we `mkScriptEnv` and check if dynamic compilation was enabled (adding a new field to `ScriptEnv` saying if we were building with dynamic or not.) This seems a lot simpler and efficient!